python/ray/experimental/channel/common.py

```diff
@@ -226,6 +228,8 @@ def read(self, timeout: Optional[float] = None) -> Any:
         zero-copy (e.g., bytes or a numpy array) *and* the object is still in scope.
 
         Args:
+            actor_id: The actor ID of the reader. This is used to determine
+                which channel to read from for CompositeChannel.
```

Comment:
we can avoid having read take in actor_id for every implementation of read by having a different read_with_actor_id function for CompositeChannel, and having readers check if it's a CompositeChannel and then use read_with_actor_id, but ya not sure which design makes sense, do we think actor_id could be useful to other channels in the future?

Reply:
Why do we need to pass in actor_id? When reader reads the channel, the reader/actor knows its ID, right?

Reply:
So the reader knows the actor that owns it now that we pass it to the reader if you're referring to like the sync/background reader

But we need to give it to the channel when reading so the CompositeChannel knows what actor is calling the read, this is to avoid using runtime context to determine the reader actor, because we need to choose whether it's a driver read, current actor read, etc.

Reply:
Passing `actor_id` to `read` is a bit weird to me. Do we have other options?

Reply:
This is the only solution I've found that really worked for all test cases because when you have driver and another actor reading from different channels you can't reliably choose which inner channel to read from with just the runtime context. It is just for CompositeChannel though because you need to choose the inner channel based on it. Open to trying out any other ideas though

Reply:
Some random ideas:

1. Maybe check which actor handle is `DAGDriverProxyActor`
2. Set readers during the preparation of `ExecutableTask`

not sure which one is better

Reply:
I still think the API is not great if the user needs to pass in the reader.

Other thoughts: when creating the channel, can we explicitly pass in the proxy actor (or use a flag to indicate), and then during read, if ray.get_runtime_context().get_actor_id() is None, we can assume it's being read by the driver, and use the proxy actor to index into channel_dict.

Reply:
Ya, so I talked to @kevin85421 offline about this, and experimenting on a separate branch with that idea. The main issue around it is that when the driver is actually an actor, then ray.get_runtime_context().get_actor_id() is not None. So trying to pass that driver actor id (if it exists and is read by driver) into composite channel on construction and then it knows it's driver read based on ```runtime_context_actor_id is None or == self.driver_actor_id```. Running into some failures but conceptually makes sense and should work